pandas/core/arrays/base.py

```diff
@@ -105,6 +104,16 @@ def __len__(self):
         # type: () -> int
         raise AbstractMethodError(self)
 
+    def __iter__(self):
+        """Iterate over elements.
+
+        This needs to be implemented so that pandas recognizes extension arrays
+        as list-like. The default implementation makes successive calls to
+        ``__getitem__``, which may be slower than necessary.
+        """
+        for i in range(len(self)):
```

Comment:
why are you not just raising AbstractMethodError?

Reply:
this is tied intimately with the boxing concept of scalars, but this should be decoupled from getitem (yes its a valid implementation, but you would never actually do it this way), rather you would iterate on some underlying value and box it, not index into and as a side effect use getitem to box.

Reply:
> rather you would iterate on some underlying value

We don't know how the subclass is storing things though, so there's no `.values` to iterate over.

Reply:
so my point is that you cannot implement a default here, make the sub-classes do it. yes it might repeat some code, but its much better there i think.

Reply:
>  you cannot 

Yes, we can (see the code). You "don't want" it, that something different :) 
But I don't really understand why the above is not a good default implementation. I think a lot subclasses will do exactly this.

Reply:
I think it still might be better raise here by default and force the subclass to implement this (even if this impl). It is SO important for subclasses to pay attention to this, it should just be required.

Reply:
You can't do much better than this implementation. 
To give you some context: in geopandas, I have a cython method that converts a GeometryArray to numpy object array of Geometry objects, which takes ca 80ms on an array of 10000 points. Simply doing `np.array([a[i] for i in range(len(a))], dtype=object)` is ca 108 ms. This small difference is because by definition in each iteration I need to create a python object.